options.txt
// Sample options.txt configuration file for Pi1541



// If you would like to change the drive number then specify it here (8 is the default)
//deviceID = 9

// If you are using the split line hardware option (ie Option B) then you need to specify this option
//splitIECLines = 1
// If you are using some type of hardware that requires the lines to be inverted then specify these options here
//invertIECInputs = 1
//If you are using a 7407
//invertIECOutputs = 0

// If you are using the Pi's composite video out then these options allow you to experiment with the display resolution
//ScreenWidth = 512
//ScreenHeight = 384

// By default Pi1541 will search the root of the SD card for a ROM file and use this for ROM1.
// The first file found with either of the following names will be used; d1541.rom, dos1541, d1541II, Jiffy.bin
// You can override the default ROM1 by specifying it here
//ROM1 = YourDefaultRom
// You can specify additional ROMS here (up to 7)
// You can then use the function keys (F1-F7) on the keyboard to change which one you would like to use (only in browse mode and not in emulation mode)
// You can also use the buttons to swap ROMs by holding down button 1 and pressing one other buttons 2-5
// .LST files can also specify one of these ROMS that will automatically be selected when the LST file is loaded.
//ROM2 = Jiffy.bin
//ROM3 = d1541II

// Reference you 1581 ROM here
//ROM1581 = 1581-rom.318045-02.bin
//ROM1581 = JiffyDOS_1581.bin

// The rate (in seconds) a long selection is scrolled
scrollHighlightRate = 0.07

// Enable RAMBoard emulation - 8k drive RAM expansion at 0x8000
//RAMBOard = 1

//ChargenFont = chargen		// 8 bit font file

// If using CBMFileBrowser then it is best to specify this option. When the computer resets the Pi will always revert back to the root folder ready to load CBMFileBrowser again.
OnResetChangeToStartingFolder = 1

// If you use FB64 (CBMFileBrowser) and want to use a fast loader cartridge (AR6, EFL, FC3) to load it then use this option to automatically mount it. 
//AutoMountImage = fb.d64 // You MUST have a disk image in \1541 with this filename

// If you are using a FB128 in 128 mode you can get FB128 to auto boot using this option
//AutoBootFB128 = 1

// If you are using a 128 you can auto boot anything using this option
// This over-rides AutoBootFB128
//128BootSectorName = bootsect.128

// If you would ever like to disable browse mode completely you can do so here
//DisableSD2IECCommands = 1

// This option displays the IEC bus activity on the bottom of the Pi's screen
GraphIEC = 1

// If you have hardware with a peizo buzzer (the type without a generator) then you can use this option to hear the head step
//SoundOnGPIO = 1
//SoundOnGPIODuration = 1000 // Length of buzz in micro seconds
//SoundOnGPIOFreq = 1200 // Frequency of buzz in Hz

// You can create 320x240 PNG files with the same name as your disk images. With this option turned on they will be displayed on the Pi's screen.
//DisplayPNGIcons = 1

// If you would like to specify what file will be loaded by LOAD"*" in browse mode then specify it here
//StarFileName = somefile

// Alt-N creates a new disk image.
// Names are formed automatically as autoname001.d64, autoname002.d64 etc...
// change the base of the filename here
//AutoBaseName = autoname

// If you are using a LCD screen then specify it here
//LCDName = ssd1306_128x64
//LCDName = ssd1306_128x32
//LCDName = sh1106_128x64
// If you are using a LCD screen and you would like PageUp and PageDown keys to work with it then specify this option
//KeyboardBrowseLCDScreen = 1

// change startup logo on oled - 1541ii or 1541classic
//LcdLogoName = 1541ii
//LcdLogoName = 1541classic
//LcdLogoName = customfile.raw

// If you are using I2C LCD you can optionally change what pins it is connected to.
// (defaults to 0 for non-split lines (Option A) or 1 for split lines (Option B))
//i2cBusMaster = 0 //SDA - pin 27 SCL - pin 28
//i2cBusMaster = 1 //SDA - pin 3 SCL - pin 5
//i2cLcdAddress = 60	// I2C display address in decimal and shifted. 60 == 0x78, 61 == 0x7A
//i2cLcdFlip = 1 // Rotate i2c LCD screen 180 degrees
//i2cLcdOnContrast = 127 // Allows you to adjust the contrast on your i2c LCD screen
//i2cScan = 1 // scan i2c bus and display addresses on screen
//i2cLcdUseCBMChar = 0 // set it to 1 to use CBM font on LCD. Small but fun !

//QuickBoot = 0		// faster startup
//ShowOptions = 0	// display some options on startup screen 
//IgnoreReset = 0

// If you are using a rotary button set this option to 1
// in this case you have to map the clk pin to buttonUp
// and dt pin to buttonDown (see below)
//useRotaryButton = 0

// You can remap the physical button functions
// numbers correspond to the standard board layout
//buttonEnter = 1
//buttonUp = 2
//buttonDown = 3
//buttonBack = 4
//buttonInsert = 5


src/options.cpp
// Pi1541 - A Commodore 1541 disk drive emulator
// Copyright(C) 2018 Stephen White
//
// This file is part of Pi1541.
// 
// Pi1541 is free software : you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Pi1541 is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with Pi1541. If not, see <http://www.gnu.org/licenses/>.

#include "options.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <ctype.h>

#define INVALID_VALUE	((unsigned) -1)

char* TextParser::GetToken(bool includeSpace)
{
	bool isSpace;

	do
	{
	} while (ParseComment());

	while (*data != '\0')
	{
		isSpace = isspace(*data);

		if (!isSpace || (includeSpace && isSpace))
			break;

		data++;
	}

	if (*data == '\0')
		return 0;

	char* pToken = data;

	while (*data != '\0')
	{
		isSpace = isspace(*data);

		if ((!includeSpace && isSpace) || (*data == '\n') || (*data == '\r'))
		{
			*data++ = '\0';
			break;
		}
		data++;
	}

	return pToken;
}

void TextParser::SkipWhiteSpace()
{
	while (*data != '\0')
	{
		if (!isspace(*data))
			break;

		data++;
	}
}

bool TextParser::ParseComment()
{
	SkipWhiteSpace();

	if (*data == '\0')
		return 0;

	if (data[0] != '/')
		return false;

	if (data[1] == '/')
	{
		// One line comment
		data += 2;

		while (*data)
		{
			if (*data == '\n')
				break;

			data++;
		}
		SkipWhiteSpace();
		return true;
	}
	else if (data[1] == '*')
	{
		// Multiline comment
		data += 2;

		while (*data)
		{
			if (*data++ == '*' && *data && *data == '/')
			{
				data++;
				break;
			}
		}
		SkipWhiteSpace();
		return true;
	}
	return false;
}


Options::Options(void)
	: TextParser()
	, deviceID(8)
	, onResetChangeToStartingFolder(0)
	, extraRAM(0)
	, RAMBOard(0)
	, disableSD2IECCommands(0)
	, supportUARTInput(0)
	, graphIEC(0)
	, quickBoot(0)
	, showOptions(0)
	, displayPNGIcons(0)
	, soundOnGPIO(0)
	, soundOnGPIODuration(1000)
	, soundOnGPIOFreq(1200)
	, invertIECInputs(0)
	, invertIECOutputs(1)
	, splitIECLines(0)
	, ignoreReset(0)
	, autoBootFB128(0)
	, lowercaseBrowseModeFilenames(0)
	, screenWidth(1024)
	, screenHeight(768)
	, i2cBusMaster(1)
	, i2cLcdAddress(0x3C)
	, i2cScan(0)
	, i2cLcdFlip(0)
	, i2cLcdOnContrast(127)
	, i2cLcdUseCBMChar(0)
	, i2cLcdModel(LCD_UNKNOWN)
	, scrollHighlightRate(0.125f)
	, keyboardBrowseLCDScreen(0)
	, useRotaryButton(0)
        , buttonEnter(1)
        , buttonUp(2)
        , buttonDown(3)
        , buttonBack(4)
        , buttonInsert(5)

{
	autoMountImageName[0] = 0;
	strcpy(ROMFontName, "chargen");
	strcpy(LcdLogoName, "1541ii");
	strcpy(autoBaseName, "autoname");
	C128BootSectorName[0] = 0;
	starFileName[0] = 0;
	ROMName[0] = 0;
	ROMNameSlot2[0] = 0;
	ROMNameSlot3[0] = 0;
	ROMNameSlot4[0] = 0;
	ROMNameSlot5[0] = 0;
	ROMNameSlot6[0] = 0;
	ROMNameSlot7[0] = 0;
	ROMNameSlot8[0] = 0;
	ROMName1581[0] = 0;
}

#define ELSE_CHECK_DECIMAL_OPTION(Name) \
	else if (strcasecmp(pOption, #Name) == 0) \
	{ \
		unsigned nValue = 0; \
		if ((nValue = GetDecimal(pValue)) != INVALID_VALUE) \
			Name = nValue; \
	}
#define ELSE_CHECK_FLOAT_OPTION(Name) \
	else if (strcasecmp(pOption, #Name) == 0) \
	{ \
		float value = 0; \
		if ((value = GetFloat(pValue)) != INVALID_VALUE) \
			Name = value; \
	}

void Options::Process(char* buffer)
{
	SetData(buffer);

	char* pOption;
	while ((pOption = GetToken()) != 0)
	{
		/*char* equals = */GetToken();
		char* pValue = GetToken();

		if ((strcasecmp(pOption, "Font") == 0) || (strcasecmp(pOption, "ChargenFont") == 0))
		{
			strncpy(ROMFontName, pValue, 255);
		}
		else if ((strcasecmp(pOption, "AutoMountImage") == 0))
		{
			strncpy(autoMountImageName, pValue, 255);
		}
		ELSE_CHECK_DECIMAL_OPTION(deviceID)
		ELSE_CHECK_DECIMAL_OPTION(onResetChangeToStartingFolder)
		ELSE_CHECK_DECIMAL_OPTION(extraRAM)
		ELSE_CHECK_DECIMAL_OPTION(RAMBOard)
		ELSE_CHECK_DECIMAL_OPTION(disableSD2IECCommands)
		ELSE_CHECK_DECIMAL_OPTION(supportUARTInput)
		ELSE_CHECK_DECIMAL_OPTION(graphIEC)
		ELSE_CHECK_DECIMAL_OPTION(quickBoot)
		ELSE_CHECK_DECIMAL_OPTION(showOptions)
		ELSE_CHECK_DECIMAL_OPTION(displayPNGIcons)
		ELSE_CHECK_DECIMAL_OPTION(soundOnGPIO)
		ELSE_CHECK_DECIMAL_OPTION(soundOnGPIODuration)
		ELSE_CHECK_DECIMAL_OPTION(soundOnGPIOFreq)
		ELSE_CHECK_DECIMAL_OPTION(invertIECInputs)
		ELSE_CHECK_DECIMAL_OPTION(invertIECOutputs)
		ELSE_CHECK_DECIMAL_OPTION(splitIECLines)
		ELSE_CHECK_DECIMAL_OPTION(ignoreReset)
		ELSE_CHECK_DECIMAL_OPTION(lowercaseBrowseModeFilenames)
		ELSE_CHECK_DECIMAL_OPTION(autoBootFB128)
		ELSE_CHECK_DECIMAL_OPTION(screenWidth)
		ELSE_CHECK_DECIMAL_OPTION(screenHeight)
		ELSE_CHECK_DECIMAL_OPTION(i2cBusMaster)
		ELSE_CHECK_DECIMAL_OPTION(i2cLcdAddress)
		ELSE_CHECK_DECIMAL_OPTION(i2cScan)
		ELSE_CHECK_DECIMAL_OPTION(i2cLcdFlip)
		ELSE_CHECK_DECIMAL_OPTION(i2cLcdOnContrast)
		ELSE_CHECK_DECIMAL_OPTION(i2cLcdDimContrast)
		ELSE_CHECK_DECIMAL_OPTION(i2cLcdDimTime)
		ELSE_CHECK_DECIMAL_OPTION(i2cLcdUseCBMChar)
		ELSE_CHECK_FLOAT_OPTION(scrollHighlightRate)
		ELSE_CHECK_DECIMAL_OPTION(keyboardBrowseLCDScreen)
		ELSE_CHECK_DECIMAL_OPTION(useRotaryButton)
		ELSE_CHECK_DECIMAL_OPTION(buttonEnter)
		ELSE_CHECK_DECIMAL_OPTION(buttonUp)
		ELSE_CHECK_DECIMAL_OPTION(buttonDown)
		ELSE_CHECK_DECIMAL_OPTION(buttonBack)
		ELSE_CHECK_DECIMAL_OPTION(buttonInsert)
		else if ((strcasecmp(pOption, "AutoBaseName") == 0))
		{
			strncpy(autoBaseName, pValue, 255);
		}
		else if ((strcasecmp(pOption, "128BootSectorName") == 0))
		{
			strncpy(C128BootSectorName, pValue, 255);
		}
		else if ((strcasecmp(pOption, "StarFileName") == 0))
		{
			strncpy(starFileName, pValue, 255);
		}
		else if ((strcasecmp(pOption, "LCDLogoName") == 0))
		{
			strncpy(LcdLogoName, pValue, 255);
		}
		else if ((strcasecmp(pOption, "LCDName") == 0))
		{
			strncpy(LCDName, pValue, 255);
			if (strcasecmp(pValue, "ssd1306_128x64") == 0)
				i2cLcdModel = LCD_1306_128x64;
			else if (strcasecmp(pValue, "ssd1306_128x32") == 0)
				i2cLcdModel = LCD_1306_128x32;
			else if (strcasecmp(pValue, "sh1106_128x64") == 0)
				i2cLcdModel = LCD_1106_128x64;
		}
		else if ((strcasecmp(pOption, "ROM1581") == 0))
		{
			strncpy(ROMName1581, pValue, 255);
		}
		else if ((strcasecmp(pOption, "ROM") == 0) || (strcasecmp(pOption, "ROM1") == 0))
		{
			strncpy(ROMName, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM2") == 0)
		{
			strncpy(ROMNameSlot2, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM3") == 0)
		{
			strncpy(ROMNameSlot3, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM4") == 0)
		{
			strncpy(ROMNameSlot4, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM5") == 0)
		{
			strncpy(ROMNameSlot5, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM6") == 0)
		{
			strncpy(ROMNameSlot6, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM7") == 0)
		{
			strncpy(ROMNameSlot7, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM8") == 0)
		{
			strncpy(ROMNameSlot8, pValue, 255);
		}
	}

	if (!SplitIECLines())
	{
		invertIECInputs = false;
		// If using non split lines then only the 1st bus master can be used (as ATN is using the 2nd)
		i2cBusMaster = 0;
	}
}

unsigned Options::GetDecimal(char* pString)
{
	if (pString == 0 || *pString == '\0')
		return 0;

	return strtol(pString, NULL, 0);
}

float Options::GetFloat(char* pString)
{
	if (pString == 0 || *pString == '\0')
		return 0;

	return atof(pString);
}

const char* Options::GetRomName(int index) const
{
	switch (index)
	{
	case 1:
		return ROMNameSlot2;
	case 2:
		return ROMNameSlot3;
	case 3:
		return ROMNameSlot4;
	case 4:
		return ROMNameSlot5;
	case 5:
		return ROMNameSlot6;
	case 6:
		return ROMNameSlot7;
	case 7:
		return ROMNameSlot8;
	}
	return ROMName;
}

const char* Options::GetRomName1581() const
{
	return ROMName1581;
}

